tests/aggregation/tests.py

```diff
@@ -1402,3 +1420,199 @@ def test_coalesced_empty_result_set(self):
+        with self.assertRaisesMessage(TypeError, msg):
+            Count('age', default=0)
+
+    def test_aggregation_default_unset(self):
+        for Aggregate in [Avg, Max, Min, StdDev, Sum, Variance]:
+            with self.subTest(Aggregate):
+                result = (
+                    Author.objects.filter(age__gt=100)
+                    .aggregate(value=Aggregate('age'))
+                )
+                self.assertIsNone(result['value'])
+
+    def test_aggregation_default_zero(self):
+        for Aggregate in [Avg, Max, Min, StdDev, Sum, Variance]:
+            with self.subTest(Aggregate):
+                result = (
+                    Author.objects.filter(age__gt=100)
+                    .aggregate(value=Aggregate('age', default=0))
+                )
+                self.assertEqual(result['value'], 0)
+
+    def test_aggregation_default_integer(self):
+        for Aggregate in [Avg, Max, Min, StdDev, Sum, Variance]:
+            with self.subTest(Aggregate):
+                result = (
+                    Author.objects.filter(age__gt=100)
+                    .aggregate(value=Aggregate('age', default=21))
+                )
+                self.assertEqual(result['value'], 21)
+
+    def test_aggregation_default_expression(self):
+        for Aggregate in [Avg, Max, Min, StdDev, Sum, Variance]:
+            with self.subTest(Aggregate):
+                result = (
+                    Author.objects.filter(age__gt=100)
+                    .aggregate(value=Aggregate('age', default=Value(5) * Value(7)))
+                )
+                self.assertEqual(result['value'], 35)
+
+    def test_aggregation_default_group_by(self):
+        qs = (
+            Publisher.objects.values('name')
+            .annotate(books=Count('book'), pages=Sum('book__pages', default=0))
+            .filter(books=0)
+        )
+        self.assertEqual(
+            list(qs),
+            [{'name': "Jonno's House of Books", 'books': 0, 'pages': 0}],
+        )
+
+    def test_aggregation_default_compound_expression(self):
+        # Scale rating to a percentage; default to 50% if no books published.
+        formula = Avg('book__rating', default=2.5) * 20.0
+        queryset = Publisher.objects.annotate(rating=formula).order_by('name')
+        self.assertQuerysetEqual(queryset, [
+            ('Apress', 85.0),
+            ("Jonno's House of Books", 50.0),
+            ('Morgan Kaufmann', 100.0),
+            ('Prentice Hall', 80.0),
+            ('Sams', 60.0),
+        ], lambda obj: (obj.name, obj.rating))
+
+    def test_aggregation_default_using_time_from_python(self):
+        expr = Min(
+            'store__friday_night_closing',
+            filter=~Q(store__name='Amazon.com'),
+            default=datetime.time(17),
+        )
+        if connection.vendor == 'mysql':
+            # Workaround for ticket-30224 for MySQL 8.0+ & MariaDB 10.2+.
+            expr.default = Cast(expr.default, TimeField())
```

Comment:
When passing the value from Python for the time/date/datetime tests we need to cast the string to the appropriate type. This works for MySQL 8.0+ and MariaDB.

We may be able to address this as part of ticket-30224 by having `Value()` explicitly wrap itself with `Cast()` when it has an `output_field` of `DateField`, `DateTimeField`, or `TimeField`. Alternatively we could only target expressions passed to `Coalesce()` if we want to narrow the scope, but that would then not address any number of other related cases.